fast_pauli/cpp/include/__summed_pauli_op.hpp

```diff
@@ -173,20 +203,166 @@ template <std::floating_point T> struct SummedPauliOp
+        // coeffs_sq(a, k) = sum_{i,j} t_aij(a, i, j) * coeffs(i, k) * coeffs(j, k)
+        // If we break it up into two steps, the intermediates will be too big, so we're
+        // just going to do it all at once
+
+        // #pragma omp parallel for collapse(2)
+        fmt::println("{} {} {} {}", pauli_strings_sq.size(), _n_operators, pauli_strings.size(), pauli_strings.size());
+        for (size_t a = 0; a < pauli_strings_sq.size(); ++a)
+        {
+            for (size_t k = 0; k < _n_operators; ++k)
+            {
+                for (size_t i = 0; i < pauli_strings.size(); ++i)
+                {
+                    for (size_t j = 0; j < pauli_strings.size(); ++j)
+                    {
+                        // fmt::println("{} {} {} {}", a, k, i, j);
+                        coeffs_sq(a, k) += t_aij(a, i, j) * coeffs(i, k) * coeffs(j, k);
+                    }
+                }
+            }
+        }
+
+        return SummedPauliOp<T>(pauli_strings_sq, coeffs_sq);
+    }
+
+    /**
+     * @brief Apply the SummedPauliOp to a set of states, mathematically
+     * \f$ \big(\sum_k \sum_i h_{ik} \mathcal{\hat{P}}_i \big) \ket{\psi_t} \f$
+     *
+     * @param new_states The output states after applying the SummedPauliOp (n_dim, n_states)
+     * @param states The input states to apply the SummedPauliOp to (n_dim, n_states)
+     */
+    void apply(Tensor<2> new_states, Tensor<2> states) const
+    {
+        apply(std::execution::seq, new_states, states);
+    }
+
+    /**
+     * @brief \copydoc SummedPauliOp::apply(Tensor<2>, Tensor<2>) const
+     *
+     * @tparam ExecutionPolicy
+     */
+    template <execution_policy ExecutionPolicy>
+    void apply(ExecutionPolicy &&, Tensor<2> new_states, Tensor<2> states) const
+    {
+        if (states.extent(0) != new_states.extent(0) || states.extent(1) != new_states.extent(1))
+        {
+            throw std::invalid_argument("new_states must have the same dimensions as states");
+        }
+        size_t const n_threads = omp_get_max_threads();
+        size_t const n_ps = n_pauli_strings();
+        size_t const n_ops = n_operators();
+        size_t const n_dim = dim();
+        size_t const n_data = states.extent(1);
+
+        std::vector<std::complex<T>> new_states_thr_raw(n_threads * n_dim * n_data);
+        std::mdspan<std::complex<T>, std::dextents<size_t, 3>> new_states_thr(new_states_thr_raw.data(), n_threads,
+                                                                              n_dim, n_data);
+
+        if constexpr (is_parallel_execution_policy_v<ExecutionPolicy>)
+        {
+
+#pragma omp parallel
+            {
+#pragma omp for schedule(static)
+                for (size_t j = 0; j < n_ps; ++j)
+                {
+                    std::complex<T> summed_coeff(0, 0);
+
+                    for (size_t k = 0; k < n_ops; ++k)
+                    {
+                        summed_coeff += coeffs(j, k);
+                    }
```

Comment:
We could precompute it in this function, but I don't think it provides a big speedup here since that will still be two nested loops over j and k. These coefficients will change every iteration of the HSM weight optimization (which will happen a lot), so I don't think we should precompute (either inside or outside the function)